PHPUnit on GitHub Actions

derrabus opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

This PR adds PHPUnit jobs for all supported PHP versions to our GitHub Actions workflow. These jobs can replace the first of our three Travis CI jobs (AKA the "php_extra" one). The idea is to have a more integrated CI experience for contributors and to be prepared for future Travis outages.

derrabus committed:
Add High Deps CI job

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Add Low Deps CI job

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Add PHPUnit CI job

Signed-off-by: Alexander M. Turek <me@derrabus.de>

Nyholm commented on .github/workflows/tests.yml:
```diff
@@ -169,3 +169,202 @@ jobs:
+
+      - name: Configure composer
+        run: |
+          SYMFONY_VERSION=$(grep -m1 SYMFONY_VERSION .travis.yml | grep -o '[0-9.x]*')
```
Should we really keep .travis.yml?

derrabus replied:
I don't want to create another place that we have to keep updated. If Travis goes away completely, we'll find another way. For now, this is consistent with other GHA jobs on this repo.

Nyholm commented on .github/workflows/tests.yml:
```diff
@@ -169,3 +169,202 @@ jobs:
+          SYMFONY_VERSION=$(grep -m1 SYMFONY_VERSION .travis.yml | grep -o '[0-9.x]*')
+          SYMFONY_REQUIRE=">=$SYMFONY_VERSION"
+          COMPOSER_HOME="$(composer config home)"
+          composer self-update
```
This is not needed as far as I know.

stof replied:
indeed. `shivammathur/setup-php` already takes care of installing an uptodate one

derrabus replied:
FYI: This line is C&P from existing GHA jobs. I agree that we should remove it everywhere.

Nyholm commented on .github/workflows/tests.yml:
```diff
@@ -169,3 +169,202 @@ jobs:
+
+            # Install dependencies
+            cd ./$1
+            composer update --no-progress --ansi 2>&1  || ok=1
```
Double space

Nyholm commented on .github/workflows/tests.yml:
```diff
@@ -169,3 +169,202 @@ jobs:
+            composer update --no-progress --ansi 2>&1  || ok=1
+
+            if [ $ok -ne 0 ]; then
+              echo "::error::$1 failed"
```
You need to do `echo ::endgroup::` before here or the group will not be closed.

Nyholm replied:
Or maybe change the logic... Only run the tests if `$ok -eq 0`?

Nyholm reviewed (commented):
Thank you for working on this.

derrabus committed:
Review changes

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Exclude memcached on PHP 8

Signed-off-by: Alexander M. Turek <me@derrabus.de>

Nyholm commented on .github/workflows/tests.yml:
```diff
@@ -169,3 +167,207 @@ jobs:
+              $PHPUNIT --colors=always --exclude-group tty,benchmark,intl-data,legacy 2>&1 || ok=1
+            fi
+
+            echo
```
oops

derrabus replied:
I've added an extra line to the output on purpose to make sure the following `::endgroup::` is really on a separate line. I've had an issue with a group not closing because some error output was not terminated with a newline.

Nyholm commented on .github/workflows/tests.yml:
```diff
@@ -169,3 +167,207 @@ jobs:
+              echo "::error::$1 failed"
+            fi
+
+            # Make the tests always pass because we don't want the build to fail (yet).
```
This comment can be removed

derrabus replied:
Of course 🙈

Nyholm commented on .github/workflows/tests.yml:
```diff
@@ -169,3 +167,207 @@ jobs:
+
+      - name: Install dependencies
+        run: |
+          echo "::group::composer update"
+          composer update --no-progress --ansi
+          echo "::endgroup::"
+          echo "::group::install phpunit"
+          ./phpunit install
+          echo "::endgroup::"
```
Nitpicking: We are inconstant with when we use quotes. Before here we only use quotes when needed. 

I suggest we always use quotes for clarity.

derrabus committed:
Review changes

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
backslash-n doesn't work…

Signed-off-by: Alexander M. Turek <me@derrabus.de>

wouterj commented on .github/workflows/tests.yml:
```diff
@@ -169,3 +169,209 @@ jobs:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          coverage: "none"
+          ini-values: "memory_limit=-1"
+          php-version: "8.0"
+          # TODO: memcached is excluded because of compatibility issues with PHP 8
+          extensions: amqp, apcu, igbinary, intl, mbstring, :memcached, mongodb, redis
+          tools: flex
+
+      - name: Configure composer
+        run: |
+          SYMFONY_VERSION=$(grep -m1 SYMFONY_VERSION .travis.yml | grep -o '[0-9.x]*')
+          COMPOSER_HOME="$(composer config home)"
+          ([ -d "$COMPOSER_HOME" ] || mkdir "$COMPOSER_HOME") && cp .github/composer-config.json "$COMPOSER_HOME/config.json"
+
+          echo COMPOSER_ROOT_VERSION=$SYMFONY_VERSION.x-dev >> $GITHUB_ENV
+          echo SYMFONY_VERSION=$SYMFONY_VERSION >> $GITHUB_ENV
```
Is it an idea to use Yaml anchors to avoid all the code duplication?

E.g.

```yaml
steps:
  - name: Checkout
    uses: actions/checkout@v2

  - name: Setup PHP
    uses: shivammathur/setup-php@v2
    with:
      <<: *php_variables
      php-version: "8.0"
      # TODO: memcached is excluded because of compatibility issues with PHP 8
      extensions: amqp, apcu, igbinary, intl, mbstring, :memcached, mongodb, redis

  - *configure_composer

  - *install_phpunit
```

derrabus replied:
Maybe. I've removed the high/low deps jobs for now. That should eliminate a lot of C&P as well.

mpdude replied:
@wouterj Do YAML anchors work in GHA? Last time I checked they did not...

derrabus committed:
Skip segfaulting tests

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Remove high/low and merge with nightly

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus commented:
Update: Since Travis is back, I have removed the high/low deps jobs. Replicating those needs more work. Furthermore, it's not that bad if we lose them again for a day or two. I've adjusted the PR description accordingly.

derrabus commented on src/Symfony/Component/VarExporter/Tests/VarExporterTest.php:
```diff
@@ -67,10 +67,13 @@ public function provideFailingSerialization()
-        $a = [null, $h];
-        $a[0] = &$a;
+        // This test segfaults on the final PHP 7.2 release
+        if (\PHP_VERSION_ID !== 70234) {
+            $a = [null, $h];
+            $a[0] = &$a;
 
-        yield [$a];
+            yield [$a];
+        }
```
I was able to reproduce the segfaults we caught on GHA on my dev machine (macOS 11.4 with PHP 7.2 from Homebrew). I don't know if we can do anything about them, but since nobody is going to patch PHP 7.2 for us anymore, I decided to skip the corresponding tests.

On Travis, we get an older version of PHP 7.2 (7.2.27). I assume that's why the segfault does not happen there.

nicolas-grekas replied:
can't we stick to 7.2.27?

derrabus replied:
We probably can (have to try out), but pinning that version just because Travis wasn't updated for some time feels a bit arbitrary, doesn't it?

nicolas-grekas replied:
That'd be fine to me: we don't test patch level versions of php here. We just need a version that is capable of running the test suite.

derrabus replied:
I tried to explicitly request `7.2.27`, but Setup PHP still seems to install the latest 7.2.

<img width="206" alt="Bildschirmfoto 2021-06-14 um 16 55 19" src="https://user-images.githubusercontent.com/1506493/121913036-58b0b780-cd31-11eb-9ea2-b2913a4444a7.png">

> We just need a version that is capable of running the test suite.

And apart from those two test cases, the version is capable. And the two cases that are excluded here, are still covered in the PHP 7.1/7.3 runs.

wouterj commented:
> Update: Since Travis is back, I have removed the high/low deps jobs. Replicating those needs more work.

Does it make sense to do the opposite as well: remove the non-high/low deps jobs from Travis? It seems a bit wasteful to run the "normal" tests twice (and it's also more sad to see more red crosses if you submit a bug :grin: )

derrabus commented:
> > Update: Since Travis is back, I have removed the high/low deps jobs. Replicating those needs more work.
> 
> Does it make sense to do the opposite as well: remove the non-high/low deps jobs from Travis? It seems a bit wasteful to run the "normal" tests twice (and it's also more sad to see more red crosses if you submit a bug 😁 )

We could probably keep the regular job, but remove the logic that runs the tests on different PHP versions. That would still speed up the Travis build and we would have a backup job in case GHA lets us down. Also, we would have a reference test run we could compare a low/high deps failure against.

nicolas-grekas commented:
> Does it make sense to do the opposite as well: remove the non-high/low deps jobs from Travis

beware that Travis runs flipped tests on non-high/low deps, we should not lose that in the process!

nicolas-grekas reviewed (changes requested):
I'm putting this on hold because despite the 2+ votes, we should not merge this until we're sure we don't weaken the coverage of the test suite.

nicolas-grekas commented:
What about merging the integration tests with the new jobs? I don't see a reason to keep them separate anymore.

derrabus commented:
> we should not merge this until we're sure we don't weaken the coverage of the test suite.

Right now, the PR only adds jobs and does not remove any. That should not weaken the test suite, quite the opposite actually.

> What about merging the integration tests with the new jobs? I don't see a reason to keep them separate anymore.

The integrations tests need more setup and additional docker containers for external dependencies like LDAP and Redis. And they are typically more fragile than the regular unit tests. Also, we probably don't want to run them for each PHP branch. Keeping them separate still makes sense to me.

May I suggest to add `integration` to the `--exclude-group` section of the new job instead?

derrabus committed:
Skip integration tests for the phpunit step

Signed-off-by: Alexander M. Turek <me@derrabus.de>

nicolas-grekas commented:
I think this very PR should remove from .travis.yml what is covered here. That will ease the review of what actually moved and what is still not moved in terms of features/coverage of the CI.

Can you check that the general setup is really the same? `stty cols 120`, and all php.ini settings?

Would it be possible to display the time each component took in the collapsed view? I'm not sure. That's useful in travis to quickly know which test suite could be optimized for speed.

About the integration tests, I'd like to merge them. I agree we don't need to run them for all php versions, so we could have a flag to tell which php versions should run them (and enable the corresponding services conditionally.)

I think we should optimize the number of jobs. More precisely, I think we should target to have only one job per php version. Some of them would run integration tests, and some should run deps=low/high.

This is how the travis test suite is designed and I think that works great, both for the planet, for coverage, for the response time of our CI, and for understandability of our test matrix (I consider that too many lines in the matrix is a barrier to contribute, unless there is a strong rationale to have that many lines.)

derrabus committed:
Patch return types

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Don't parse travis.yml

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Run tests with SIGCHLD enabled PHP

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Run TTY tests as well

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Remove php_extra run from Travis

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Run TTY tests separately

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Remove TTY tests again

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus commented on .github/workflows/tests.yml:
```diff
@@ -115,57 +122,115 @@ jobs:
+        run: |
+          mkdir build
+          cd build
+          wget https://github.com/derrabus/php-pcntl-binaries/raw/main/php-7.1.3-pcntl-sigchild.tar.bz2
```
We should probably upload these binaries somewhere else. Ideas?

derrabus committed:
Add some INI settings from Travis config

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus committed:
Enable APC on CLI

derrabus committed:
Silence wget

Signed-off-by: Alexander M. Turek <me@derrabus.de>

derrabus commented:
Please have another look. I've removed the php_extra run from Travis and tried to added everything it does to my GHA setup.

* I couldn't get the TTY tests to work. I might need some help here.
* I've applied all php.ini settings from the Travis setup with the exception of `opcache.enable_cli = 1`. That setting causes the PHP 8.0 run to segfault. 😕 
* The PHPUnit bridge is not tested yet.

derrabus committed:
Fix extensions env var

Signed-off-by: Alexander M. Turek <me@derrabus.de>

nicolas-grekas commented:
Continued in #41847, thanks @derrabus for pushing this!
